### PR TITLE
Jormun: INIT - Manage best boarding positions

### PR DESCRIPTION
**Pull request by kadhikari**
* Get best_boarding_positions directory from the config: BEST_BOARDING_POSITIONS_DIR 
* Load a file with name = <coverage name>.csv if exists in the directory and associate to the instance: instance.best_boarding_positions = [..]
* Update all the PT sections before transfer or (street_network and with vias) with best boarding positions if  best_boarding_positions exist for this instance(coverage).
* Some unit tests added

For details:   https://navitia.atlassian.net/browse/NAV-1932

### CODE REVIEW

File: source/jormungandr/jormungandr/scenarios/new_default.py [hunk 1]
```diff
@@ -313,6 +314,17 @@ def fill_air_pollutants(pb_resp, instance, request_id):
                 s.air_pollutants.values.CopyFrom(pollutants_values)
 
 
+def update_best_boarding_positions(pb_resp, instance):
+    if not instance.best_boarding_positions:
+        return
+    for j in pb_resp.journeys:
+        for i in range(len(j.sections)):
+            if i > 0 and j.sections[i - 1].type == response_pb2.PUBLIC_TRANSPORT:
+                boarding_positions = get_best_boarding_position(j.sections[i], instance.best_boarding_positions)
```
Comment (xlqian) [inline source/jormungandr/jormungandr/scenarios/new_default.py]: can you define `get_best_boarding_position` as a method in `instance`?

File: source/jormungandr/jormungandr/scenarios/new_default.py [hunk 2]
```diff
@@ -313,6 +314,17 @@ def fill_air_pollutants(pb_resp, instance, request_id):
                 s.air_pollutants.values.CopyFrom(pollutants_values)
 
 
+def update_best_boarding_positions(pb_resp, instance):
+    if not instance.best_boarding_positions:
+        return
+    for j in pb_resp.journeys:
+        for i in range(len(j.sections)):
```
Comment (xlqian) [inline source/jormungandr/jormungandr/scenarios/new_default.py]:
```suggestion
        for idx, section in enumerate(j.sections)):
```

File: source/jormungandr/jormungandr/instance_manager.py [hunk 3]
```diff
@@ -123,6 +124,11 @@ def register_instance(self, config):
             config.get('pt_planners', {}),
             config.get('ghost_words', []),
         )
+        # The best_boarding_positions files if any
+        # best_boarding_positions_dir = "/home/krishna/projet/debug/positionnement_NAV-1932/support"
```
Comment (xlqian) [inline source/jormungandr/jormungandr/instance_manager.py]: remove?

File: source/jormungandr/jormungandr/scenarios/helpers.py [hunk 4]
```diff
@@ -60,3 +60,17 @@ def get_pollutants_value(distance):
...
+        return response_pb2.BoardingPosition.Value(bp_str)
+    except ValueError:
+        return None
+
+
+def fill_best_boarding_position(pb_section, boarding_positions):
+    for bp in boarding_positions:
+        pb_boarding_position_type = get_boarding_position_type(bp)
+        if pb_boarding_position_type:
+            pb_section.best_boarding_positions.append(pb_boarding_position_type)
```
Comment (xlqian) [inline source/jormungandr/jormungandr/scenarios/helpers.py]:
```suggestion
    best_positions = (get_boarding_position_type(bp) for bp in boarding_positions)
    pb_section.best_boarding_positions.extend((p in best_positions if p is not None))
```

File: source/jormungandr/jormungandr/utils.py [hunk 5]
```diff
@@ -907,3 +910,37 @@ def get_weekday(timestamp):
 
 def is_stop_point(uri):
     return uri.startswith("stop_point") if uri else False
+
+
+def read_best_boarding_positions(file_path):
+    if not os.path.exists(file_path):
+        return None
+    try:
+        my_dict = dict()
```
Comment (xlqian) [inline source/jormungandr/jormungandr/utils.py]: you can use `defaultdict`

File: source/jormungandr/jormungandr/utils.py [hunk 6]
```diff
@@ -907,3 +910,37 @@ def get_weekday(timestamp):
...
+
+
+def get_best_boarding_position(section, positions_map):
+    if not positions_map:
+        return None
+    if section.type == response_pb2.TRANSFER:
+        key = BEST_BOARDING_POSITION_KEY.format(section.origin.uri, section.destination.uri)
+        return positions_map.get(key, None)
+    elif section.type == response_pb2.STREET_NETWORK and hasattr(section, 'vias') and section.vias:
+        key = BEST_BOARDING_POSITION_KEY.format(section.origin.uri, section.vias[0].uri)
```
Comment (pbench) [inline source/jormungandr/jormungandr/utils.py]: be careful about `vias[0]`, the public transit access point may not be at position 0 in `vias`. For example I think there may be a [poi_accesss_point](https://github.com/hove-io/navitia/pull/3978) at `vias[0]` and a pt_access_point at `vias[1]` ?
Poke @azime

Comment (kadhikari) [reply source/jormungandr/jormungandr/utils.py]: I modified the code to iterate all vias and calculate best boarding position.

**Review by pbench — APPROVED** [review]

File: source/jormungandr/jormungandr/utils.py [hunk 7]
```diff
@@ -907,3 +910,26 @@ def get_weekday(timestamp):
...
+        fieldnames = ['from_id', 'to_id', 'positionnement_navitia']
+        with open(file_path) as f:
+            csv_reader = csv.DictReader(f, fieldnames)
+            # skip the header
+            next(csv_reader)
+
+            for line in csv_reader:
+                key = BEST_BOARDING_POSITION_KEY.format(line['from_id'], line['to_id'])
+                value = line['positionnement_navitia']
+                my_dict.setdefault(key, []).append(value)
```
Comment (xlqian) [inline source/jormungandr/jormungandr/utils.py]: use enum

File: source/jormungandr/jormungandr/instance_manager.py [hunk 8]
```diff
@@ -123,6 +124,10 @@ def register_instance(self, config):
             config.get('pt_planners', {}),
             config.get('ghost_words', []),
         )
+        # The best_boarding_positions files if any
+        best_boarding_positions_dir = app.config.get(str('BEST_BOARDING_POSITIONS_DIR'), None)
+        file_path = '{}/{}.csv'.format(best_boarding_positions_dir, instance.name)
```
Comment (xlqian) [inline source/jormungandr/jormungandr/instance_manager.py]: use path join

**Review by azime — APPROVED** [review]

**Review by pbench — APPROVED** [review]

**Pull request merged**